src/symbol/collision_index.js

```diff
@@ -143,99 +124,119 @@ class CollisionIndex {
             symbol,
             lineVertexArray,
             labelPlaneMatrix,
-            projectionCache,
-            /*return tile distance*/ true);
+            projectionCache);
 
         let collisionDetected = false;
         let inGrid = false;
         let entirelyOffscreen = true;
 
-        const tileToViewport = projectedAnchor.perspectiveRatio * textPixelRatio;
-        // pixelsToTileUnits is used for translating line geometry to tile units
-        // ... so we care about 'scale' but not 'perspectiveRatio'
-        // equivalent to pixel_to_tile_units
-        const pixelsToTileUnits = 1 / (textPixelRatio * scale);
-
-        let firstTileDistance = 0, lastTileDistance = 0;
         if (firstAndLastGlyph) {
-            firstTileDistance = this.approximateTileDistance(firstAndLastGlyph.first.tileDistance, firstAndLastGlyph.first.angle, pixelsToTileUnits, projectedAnchor.cameraDistance, pitchWithMap);
-            lastTileDistance = this.approximateTileDistance(firstAndLastGlyph.last.tileDistance, firstAndLastGlyph.last.angle, pixelsToTileUnits, projectedAnchor.cameraDistance, pitchWithMap);
-        }
-
-        for (let k = 0; k < collisionCircles.length; k += 5) {
-            const anchorPointX = collisionCircles[k];
-            const anchorPointY = collisionCircles[k + 1];
-            const tileUnitRadius = collisionCircles[k + 2];
-            const boxSignedDistanceFromAnchor = collisionCircles[k + 3];
-            if (!firstAndLastGlyph ||
-                (boxSignedDistanceFromAnchor < -firstTileDistance) ||
-                (boxSignedDistanceFromAnchor > lastTileDistance)) {
-                // The label either doesn't fit on its line or we
-                // don't need to use this circle because the label
-                // doesn't extend this far. Either way, mark the circle unused.
-                markCollisionCircleUsed(collisionCircles, k, false);
-                continue;
+            const radius = circlePixelDiameter * 0.5 * perspectiveRatio + textPixelPadding;
+            const screenPlaneMin = new Point(-viewportPadding, -viewportPadding);
+            const screenPlaneMax = new Point(this.screenRightBoundary, this.screenBottomBoundary);
+            const interpolator = new PathInterpolator();
+
+            // Construct a projected path from projected line vertices. Anchor points are ignored and removed
+            const first = firstAndLastGlyph.first;
+            const last = firstAndLastGlyph.last;
+
+            let projectedPath = first.path.slice(1).reverse().concat(last.path.slice(1));
```

Comment:
@mourner would Array.push be faster than concat?

Reply:
Yeah — if this is on a hot path, I would rewrite this with a simple `for` loop to avoid the 3 array allocations.

Reply:
nice catch! I'll try to get rid of the extra array allocations